x-pack/plugins/security/server/config.ts

```diff
@@ -5,50 +5,90 @@
  */
 
 import crypto from 'crypto';
-import { Observable } from 'rxjs';
-import { map } from 'rxjs/operators';
-import { schema, TypeOf } from '@kbn/config-schema';
-import { PluginInitializerContext } from '../../../../src/core/server';
+import { schema, Type, TypeOf } from '@kbn/config-schema';
+import { Logger } from '../../../../src/core/server';
 
-export type ConfigType = ReturnType<typeof createConfig$> extends Observable<infer P>
-  ? P
-  : ReturnType<typeof createConfig$>;
+export type ConfigType = ReturnType<typeof createConfig>;
 
-function getCommonProviderSchemaProperties(providerType: string) {
+const providerOptionsSchema = (providerType: string, optionsSchema: Type<any>) =>
+  schema.conditional(
+    schema.siblingRef('providers'),
+    schema.arrayOf(schema.string(), {
+      validate: providers => (!providers.includes(providerType) ? 'error' : undefined),
+    }),
+    optionsSchema,
+    schema.never()
+  );
+
+type ProvidersCommonConfigType = Record<
+  'enabled' | 'showInSelector' | 'order' | 'description',
+  Type<any>
+>;
+function getCommonProviderSchemaProperties(overrides: Partial<ProvidersCommonConfigType> = {}) {
   return {
     enabled: schema.boolean({ defaultValue: true }),
+    showInSelector: schema.boolean({ defaultValue: true }),
     order: schema.number({ min: 0 }),
-    description: schema.string({ defaultValue: providerType }),
+    description: schema.maybe(schema.string()),
+    ...overrides,
   };
 }
 
-function getUniqueProviderSchema(providerType: string) {
+function getUniqueProviderSchema(
+  providerType: string,
+  overrides?: Partial<ProvidersCommonConfigType>
+) {
   return schema.maybe(
-    schema.recordOf(
-      schema.string(),
-      schema.object(getCommonProviderSchemaProperties(providerType)),
-      {
-        validate(config) {
-          if (Object.values(config).filter(provider => provider.enabled).length > 1) {
-            return `Only one "${providerType}" provider can be configured.`;
-          }
-        },
-      }
-    )
+    schema.recordOf(schema.string(), schema.object(getCommonProviderSchemaProperties(overrides)), {
+      validate(config) {
+        if (Object.values(config).filter(provider => provider.enabled).length > 1) {
+          return `Only one "${providerType}" provider can be configured.`;
+        }
+      },
+    })
   );
 }
 
+type ProvidersConfigType = TypeOf<typeof providersConfigSchema>;
 const providersConfigSchema = schema.object(
   {
-    basic: getUniqueProviderSchema('basic'),
-    token: getUniqueProviderSchema('token'),
+    basic: getUniqueProviderSchema('basic', {
+      description: schema.maybe(
+        schema.any({
+          validate: () => '`basic` provider does not support custom description.',
+        })
+      ),
+      showInSelector: schema.boolean({
+        defaultValue: true,
+        validate: value => {
+          if (!value) {
+            return '`basic` provider only supports `true` in `showInSelector`.';
```

Comment:
> Is there any harm in allowing users to hide the basic authentication login form, and still use the Authorization: Basic xxx headers?

Without login form users won't be able to log in (=create session) with `basic` (unlike providers that support IdP initiated login and such), so the only role of `basic` in this case will be to indirectly influence BWC behavior of `http` (assuming `autoSchemesEnabled` is set). That sounds a bit awkward and not obvious to me to be honest.

To support this use case I'd rather expect users to not enable `basic` at all and tweak `http` instead so that the intention is clear:
```yaml
xpack.security.authc.http.schemes: [apikey, basic]
```
What do you think?

Reply:
I keep forgetting that `http` is configured separately... 

>To support this use case I'd rather expect users to not enable basic at all and tweak http instead so that the intention is clear:
>
> `xpack.security.authc.http.schemes: [apikey, basic]`
> What do you think?

That makes sense to me.